contracts/PrepaidCardMarket_v2.sol

```diff
@@ -0,0 +1,304 @@
+    address prepaidCard,
+    address customer,
+    bytes32 sku,
+    uint256 askPrice
+  );
+  event TrustedProvisionerAdded(address token);
+  event TrustedProvisionerRemoved(address token);
+  event PausedToggled(bool paused);
+
+  function initialize(address owner) public override initializer {
+    paused = false;
+    OwnableInitialize(owner);
+  }
+
+  function setup(
+    address _exchangeAddress,
+    address _prepaidCardManagerAddress,
+    address _provisioner,
+    address _tokenManager,
+    address[] calldata _trustedProvisioners,
+    address _versionManager
+  ) external onlyOwner {
+    exchangeAddress = _exchangeAddress;
+    prepaidCardManagerAddress = _prepaidCardManagerAddress;
+    provisioner = _provisioner;
+    tokenManager = _tokenManager;
+    versionManager = _versionManager;
+
+    for (uint256 i = 0; i < _trustedProvisioners.length; i++) {
+      _addTrustedProvisioner(_trustedProvisioners[i]);
+    }
+
+    emit Setup();
+  }
+
+  function setPaused(bool _paused) external onlyOwner {
+    paused = _paused;
+    emit PausedToggled(_paused);
+  }
+
+  function getTrustedProvisioners() external view returns (address[] memory) {
+    return trustedProvisioners.values();
+  }
+
+  function _addTrustedProvisioner(address _token) internal {
+    trustedProvisioners.add(_token);
+    emit TrustedProvisionerAdded(_token);
+  }
+
+  function removeTrustedProvisioner(address _token) external onlyOwner {
+    trustedProvisioners.remove(_token);
+    emit TrustedProvisionerRemoved(_token);
+  }
+
+  function provisionPrepaidCard(address customer, bytes32 sku)
+    external
+    returns (bool)
+  {
+    require(!paused, "Contract is paused");
+    require(
+      trustedProvisioners.contains(msg.sender),
+      "Only trusted provisioners allowed"
+    );
+    require(asks[sku] > 0, "Can't provision SKU with 0 askPrice");
+
+    PrepaidCardManager prepaidCardManager = PrepaidCardManager(
+      prepaidCardManagerAddress
+    );
+
+    address token = skus[sku].issuingToken;
+    uint256 faceValue = skus[sku].faceValue;
+    string memory customizationDID = skus[sku].customizationDID;
+    uint256 priceToCreatePrepaidCard = prepaidCardManager.priceForFaceValue(
+      token,
+      faceValue
+    ); // in Wei
+
+    uint256[] memory issuingTokenAmounts = new uint256[](1);
+    uint256[] memory spendAmounts = new uint256[](1);
+
+    issuingTokenAmounts[0] = Exchange(exchangeAddress).convertFromSpend(
+      token,
+      faceValue
+    );
+
+    spendAmounts[0] = faceValue;
+
+    address issuer = skus[sku].issuer;
+    address issuerSafe = skus[sku].issuerSafe;
+
+    IERC677(token).safeTransferAndCall(
+      prepaidCardManagerAddress,
+      priceToCreatePrepaidCard,
+      abi.encode(
+        customer,
+        issuingTokenAmounts,
+        spendAmounts,
+        customizationDID,
+        address(0), // marketAddress - we probably don't need it in this case?
+        issuer,
+        issuerSafe
+      )
+    );
```

Comment:
we should require() that this command is successful. the `safeTransferAndCall()` returns a boolen success if the transfer and call was able to happen. We whould wrap this entire call in `require()`

Reply:
or actually, I think the point of _safe_TransferAndCall, the thing that makes it "safe", is that it wraps this for you. so never mind.